File: android/src/main/java/com/audioanalyzer/AudioAnalyzerModule.java
package com.audioanalyzer;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.WritableArray;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.WritableNativeArray;
import com.facebook.react.bridge.WritableNativeMap;
import com.facebook.react.module.annotations.ReactModule;

@ReactModule(name = AudioAnalyzerModule.NAME)
public class AudioAnalyzerModule extends ReactContextBaseJavaModule {
  public static final String NAME = "AudioAnalyzer";

  public AudioAnalyzerModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  @Override
  @NonNull
  public String getName() {
    return NAME;
  }

  static {
    System.loadLibrary("cpp");
  }

  private static native AmplitudeData[] analyzeAudio(String filepath, double groupBySeconds);

  @ReactMethod
  public void analyzeAudio(String filepath, double groupBySeconds, Promise promise) {
    try {
      AmplitudeData[] analyze = analyzeAudio(filepath, groupBySeconds);

      WritableArray resultArray = new WritableNativeArray();
      for (AmplitudeData data : analyze) {
        WritableMap dataMap = new WritableNativeMap();
        dataMap.putDouble("timeInSeconds", data.getTimeInSeconds());
        dataMap.putDouble("amplitude", data.getAmplitude());
        resultArray.pushMap(dataMap);
      }
      promise.resolve(resultArray);
    } catch (Exception e) {
      promise.reject(e);
    }
  }
}

File: android/src/main/java/com/audioanalyzer/AudioAnalyzerModule.kt
package com.audioanalyzer

import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.turbomodule.core.interfaces.CallInvokerHolder

class AudioAnalyzerModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName() = NAME
    companion object {
        const val NAME = "AudioAnalyzer"
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun install(): Boolean {
        return try {
            System.loadLibrary("simple")

            this.reactApplicationContext.javaScriptContextHolder?.let { contextHolder ->
                this.reactApplicationContext.catalystInstance.jsCallInvokerHolder?.let { callInvokerHolder: CallInvokerHolder ->
                    this.nativeInstall(contextHolder.get(), callInvokerHolder)

                    Log.i(NAME, "Installed TestApp \uD83E\uDD84!")

                    return true
                }
            }

            false
        } catch (e: Exception) {
            return false
        }
    }

    private external fun nativeInstall(jsi: Long, callInvoker: CallInvokerHolder)
}
